src/tensor_ops/upscale2d/mod.rs

```diff
@@ -0,0 +1,210 @@
+mod cpu_kernel;
+
+#[cfg(feature = "cuda")]
+mod cuda_kernel;
+
+use crate::{
+    shapes::*,
+    tensor::{DeviceStorage, HasErr, PutTape, SplitTape, Tape, Tensor, ZerosTensor},
+};
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct Upscale2DOp {
+    pub batch: usize,
+    pub chan: usize,
+    pub h_in: usize,
+    pub h_out: usize,
+    pub w_in: usize,
+    pub w_out: usize,
+}
+
+impl Upscale2DOp {
+    fn new([b, c, h_in, w_in]: [usize; 4], [h_out, w_out]: [usize; 2]) -> Self {
+        assert!(h_out>=h_in, "Output height must be larger than input height");
+        assert!(w_out>=w_in, "Output width must be larger than input width");
+        Self {
+            batch: b,
+            chan: c,
+            h_in,
+            h_out,
+            w_in,
+            w_out,
+        }
+    }
+}
+
+macro_rules! upscale2d {
+    (Kernel=$Kernel:ident, ConstTrait=$ConstTrait:ident, TryTrait=$TryTrait:ident, Meth=$Meth:ident, TryMeth=$TryMeth:ident) => {
+        pub trait $Kernel<E: Unit>: DeviceStorage {
+            fn forward<I: Shape, O: Shape>(
+                &self,
+                op: Upscale2DOp,
+                inp: &Tensor<I, E, Self>,
+                out: &mut Tensor<O, E, Self>,
+            ) -> Result<(), Self::Err>;
+
+            fn backward<I: Shape, O: Shape>(
+                &self,
+                op: Upscale2DOp,
+                inp: &Tensor<I, E, Self>,
+                grad_inp: &mut Self::Vec<E>,
+                out: &Tensor<O, E, Self>,
+                grad_out: &Self::Vec<E>,
+            ) -> Result<(), Self::Err>;
+        }
+
+        pub trait $ConstTrait<const OH: usize, const OW: usize>: HasErr {
+            type Output;
+            fn try_upscale2d(self) -> Result<Self::Output, Self::Err>;
+        }
+
+        pub trait $TryTrait {
+            fn $Meth<const OH: usize, const OW: usize>(self) -> Self::Output
+            where
+                Self: $ConstTrait<OH, OW>,
+            {
+                self.try_upscale2d().unwrap()
+            }
+            fn $TryMeth<const OH: usize, const OW: usize>(
+                self,
+            ) -> Result<Self::Output, Self::Err>
+            where
+                Self: $ConstTrait<OH, OW>,
+            {
+                self.try_upscale2d()
+            }
+        }
+        impl<T> $TryTrait for T {}
+
+        impl<
+                C: Dim,
+                const H: usize,
+                const W: usize,
+                const OH: usize,
+                const OW: usize,
```

Comment:
I'm actually thinking we don't need const dimensions for upscale. I think we can take the approach we have with [broadcast](https://github.com/coreylowman/dfdx/blob/main/src/tensor_ops/broadcast_to.rs#L17), where we do:

```rust
trait Upscale<E> {
    fn upscale2d<OH: ConstDim, OW: ConstDim>(self) -> ...;
    fn try_upscale2d<OH: ConstDim, OW: ConstDim>(self) -> ...;
    fn upscale2d_like<OH: Dim, OW: Dim>(self, size: (OH, OW)) -> ...;
    fn try_upscale2d_like<OH: Dim, OW: Dim>(self, size: (OH, OW)) -> ...;
}
```

Reply:
For the trait yeah, the module still will though.